### PR TITLE
Add Workspace Events to NodeServices

### PR DESCRIPTION
**Pull request by ikeough**
This PR addresses [MAGN-5461](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5461).

## How
- Adds a `Guid` property to `WorkspaceModel` which is required for unique identification of `WorkspaceModel` objects.
- Implements static methods on `DynamoServices` for `WorkspaceAdded`, and `WorkspaceRemoved`, which pass event arguments containing the `Name` and the `Guid` of the modified workspace.
- Adds system tests to verify that the events are properly triggered.
- Renames `NodeServices.dll` to `DynamoServices.dll`.

PTAL: 

- [ ] @lukechurch

### CODE REVIEW

File: src/Libraries/Samples/SampleIntegration/TracedHog.cs [hunk 1]
```diff
@@ -69,7 +71,7 @@ public HogID(SerializationInfo info, StreamingContext context)
 
 
 [DSNodeServices.RegisterForTrace]
-public class TracedHog
+public class TracedHog : IDisposable
 {
```
Comment (lukechurch) [inline src/Libraries/Samples/SampleIntegration/TracedHog.cs]: The document is written to be very incremental in adding features one at a time to the class. So I think we might consider a 'TracedLifeCycleManagedHog' for adding dispose services.

What do you think?

**Comment by lukechurch** [comment]
Do we have a description of the use cases that this is intended to solve? This allows the nodes to depend on the UI behaviour of Dynamo. I'm not saying that's never a good idea, but it does create a slightly concerning coupling.

Lets chat about the use cases to see if there is a different way of handling it with regards to the execution of the graph, rather than an interaction with the UI?

**Comment by ikeough** [comment]
The case this is meant to solve is, as outlined in MAGN-5461, the library developer needs a way to know when a workspace is opened or closed. We can just as well make these events fire when the workspaces collection is changed, which would avoid any notion of UI interaction, and would just telegraph that a workspace had been added or removed.

File: src/NodeServices/WorkspaceEvents.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+
+namespace DSNodeServices
```
Comment (Benglin) [inline src/NodeServices/WorkspaceEvents.cs]: Perhaps it's just me, but I'd prefer to see a `Dynamo.Services` namespace where we put all the services from the future. For your case, it isn't even about *nodes* to begin with.

Comment (ikeough) [inline src/NodeServices/WorkspaceEvents.cs]: That's a good point Ben. I had my blinders on about the user story having to do with accessing this from node libraries, but Dynamo.Services make more sense.

Comment (ikeough) [inline src/NodeServices/WorkspaceEvents.cs]: NodeServices has been renamed to DynamoServices.

**Comment by ikeough** [comment]
@lukechurch ping!

**Comment by lukechurch** [comment]
@ikeough Sorry didn't realise this was needing me to look again

File: src/DynamoCore/Models/WorkspaceModel.cs [hunk 3]
```diff
@@ -39,6 +39,7 @@ public abstract class WorkspaceModel : NotificationObject, ILocatable, IUndoRedo
         private readonly ObservableCollection<NodeModel> nodes;
         private readonly ObservableCollection<NoteModel> notes;
         private readonly UndoRedoRecorder undoRecorder;
+        private Guid guid;
```
Comment (lukechurch) [inline src/DynamoCore/Models/WorkspaceModel.cs]: I really don't like relying on the default struct initilizer code for Guids. I'm not even sure if it's Guid.Empty - in which case this won't work.

Could we add a call in the ctor of the model guid = Guid.NewGuid() or something similar?

File: src/Libraries/Samples/SampleIntegration/TracedHog.cs [hunk 4]
```diff
@@ -92,13 +94,20 @@ private TracedHog(double x, double y, int id)
         this.ID = id;
 
         TracedHogManager.RegisterHogForID(id, this);
+
+        WorkspaceEvents.WorkspaceAdded += WorkspaceEventsWorkspaceAdded;
+    }
+
+    void WorkspaceEventsWorkspaceAdded(WorkspacesModificationEventArgs args)
```
Comment (lukechurch) [inline src/Libraries/Samples/SampleIntegration/TracedHog.cs]: Could you add a section to the Host integration guide on the wiki of "events and when to use them"? If you don't have the cycles, let me know and I'll try and do it.

**Comment by lukechurch** [comment]
lgtm after comment about guids is addressed

File: src/DynamoCore/Models/WorkspaceModel.cs (hunk 3 above)
Comment (ikeough) [inline src/DynamoCore/Models/WorkspaceModel.cs]: Good catch. Thanks @lukechurch. Adding it now.

**Pull request merged**